### PR TITLE
Fix mspi post RFC

### LINKED ISSUE #74349: CMake warning because of new MSPI support
**Describe the bug**

Kconfig leads to a wrong data base for CMake since commit https://github.com/zephyrproject-rtos/zephyr/commit/eccc2ba8c47591e8ab0fced054a2f3e26e2360e3:

https://github.com/zephyrproject-rtos/zephyr/blob/66630961627f4d8625e2dbe1cb7afe4abd690b2e/drivers/flash/Kconfig.mspi#L4-L9

As soon as the high-level symbol `FLASH` is activated, which is a common procedure at application level, an MSPI driver is automatically activated. However, this is now only available for a single instance (MSPI ATXP032 driver). This means that the drivers for the SoC flash controllers can still be built, but unfortunately only with an extremely unattractive CMake warning:

```
CMake Warning at /.../workspace/zephyr/CMakeLists.txt:915 (message):
  No SOURCES given to Zephyr library: drivers__mspi

  Excluding target from build.
```

**To Reproduce**

1. properly arranged Zephyr West workspace
2. `west build -b nucleo_f746zg -p always -d build/flash_shell zephyr/samples/drivers/flash_shell`
3. See warning above

**Expected behavior**

Application `zephyr/samples/drivers/flash_shell` will build without any warnings.

**Impact**

Showstopper, a really good QA pipelines leads build warnings to errors.

**Logs and console output**

Non.

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain Zephyr SDK 0.16.5
 - latest Zephyr main branch (v3.7.0-rc1)

**Additional context**

Non.

### PR DESCRIPTION
**Pull request by Swift-T**
This PR fixes https://github.com/zephyrproject-rtos/zephyr/issues/74349
Also fixes issue mentioned https://github.com/zephyrproject-rtos/zephyr/pull/71769#discussion_r1640567801 to make MSPI device bindings independent of SoC.

### CODE REVIEW

**Review by RichardSWheatley — DISMISSED** [review]
Nice work!

**Comment by rexut** [comment]
LGTM, please merge in ASAP

**Review by AlessandroLuo — DISMISSED** [review]

**Review by swift-tk — DISMISSED** [review]

**Comment by swift-tk** [comment]
@nashif Please help merge. Thanks.

File: drivers/flash/Kconfig.mspi
Comment (pdgendt) [inline drivers/flash/Kconfig.mspi:8]: I really dislike the `driver (devictree) > subsystem` dependency, perhaps introducing `config FLASH_MSPI_DRIVER_ENABLED` keeps it cleaner.

But why can't we make `FLASH_MSPI` an option that needs to be enabled explicitly?

**Review by pdgendt — CHANGES_REQUESTED** [review]
NACK for now.

Also the `drivers: fix Kconfig.mspi` commit subject should have something more meaningful.

File: drivers/flash/Kconfig.mspi
Comment (swift-tk) [reply drivers/flash/Kconfig.mspi:8]: > I really dislike the `driver (devictree) > subsystem` dependency, perhaps introducing `config FLASH_MSPI_DRIVER_ENABLED` keeps it cleaner.
> 
> But why can't we make `FLASH_MSPI` an option that needs to be enabled explicitly?

I'd hate having all that DT_HAS_* macros there as it will look very messy as the list of drivers grow.
I was thinking if there is anyway to put a generated Kconfig there based on the common include binding. In my case, it is mspi-device.yaml. 
I don't think currently there is anyway to do that without making the DTS compatible string a bit more messy. i.e. "ambiq,mspi-device", "mspi-aps6404l", "mspi-device"; Then I could put a single DT_HAS_MSPI_DEVICE_ENABLED there. But I think it is kind of an overkill.

What do you mean by driver (devictree) > subsystem dependency? I thought this is supposedly the correct way. The CONFIG_FLASH=y is always explicit. FLASH_MSPI still depends on FLASH even if I deleted it there as the Kconfig.mspi is included within if FLASH in https://github.com/zephyrproject-rtos/zephyr/blob/4d0a8c1e8b8316ec5071a3fcc98cd292d3bc4e6d/drivers/flash/Kconfig#L186

Comment (pdgendt) [reply drivers/flash/Kconfig.mspi:8]: I would remove `default y` from `config FLASH_MSPI` and add these explicitly to boards that need it.

Comment (pdgendt) [reply drivers/flash/Kconfig.mspi:8]: Another option is to drop `config FLASH_MSPI` and let the drivers depend on `config FLASH` directly.

The `config FLASH_MSPI` does not impact the build whatsoever except allowing the two current drivers. Or would this change in the future? AFAICS they just implement flash APIs.

Comment (swift-tk) [reply drivers/flash/Kconfig.mspi:8]: Similar thing has been done in Kconfig.mcux. This allows us not to repeatly write common things for all its members i.e. FLASH_MSPI_EMUL_DEVICE and FLASH_MSPI_ATXP032. So I think keeping `config FLASH_MSPI` is much cleaner.
The `config FLASH_MSPI` currently generate cmake warning for non-compatible platforms. I'd say it is a temp fix until we could figure out how to generate a macro for common include bindings without flooding the output .config file.

Comment (swift-tk) [reply drivers/flash/Kconfig.mspi:8]: Having CONFIG_FLASH_MSPI across boards, examples and tests seems redundant. 

> Drivers on these busses should use select to allow the automatic activation of the necessary bus drivers when devices on the bus are enabled in the devicetree.
from https://docs.zephyrproject.org/latest/build/kconfig/tips.html#select-recommendations

Comment (pdgendt) [reply drivers/flash/Kconfig.mspi:8]: You left out the important part:
> An exemption are busses like I2C and SPI, and following the same thought process things like MFD as well.

And you select the bus as is documented with `select MSPI`.

> The config FLASH_MSPI currently generate cmake warning for non-compatible platforms. I'd say it is a temp fix until we could figure out how to generate a macro for common include bindings without flooding the output .config file.

Let's make it a hidden option then, that is selected by the drivers:

```
# Copyright (c) 2024 Ambiq Micro Inc. <www.ambiq.com>
# SPDX-License-Identifier: Apache-2.0

menu "MSPI flash driver"

config FLASH_MSPI
	bool
	select FLASH_HAS_DRIVER_ENABLED
	select MSPI
	help
	  MSPI flash drivers are enabled.

config FLASH_MSPI_EMUL_DEVICE
	bool "MSPI flash device emulator"
	default y
	depends on DT_HAS_ZEPHYR_MSPI_EMUL_FLASH_ENABLED
	select FLASH_MSPI
	select FLASH_HAS_PAGE_LAYOUT
	select FLASH_HAS_EXPLICIT_ERASE

config FLASH_MSPI_ATXP032
	bool "MSPI ATXP032 driver"
	default y
	depends on DT_HAS_MSPI_ATXP032_ENABLED
	select FLASH_MSPI
	select FLASH_HAS_PAGE_LAYOUT
	select FLASH_HAS_EXPLICIT_ERASE
	select FLASH_JESD216
	select MSPI_AMBIQ_AP3 if SOC_SERIES_APOLLO3X

endmenu

```

`depends on FLASH` is redundant because of `flash/Kconfig`

Comment (swift-tk) [reply drivers/flash/Kconfig.mspi:8]: Well that changed the hierarchy and looks a bit weird. It is not perfect either way.

Comment (pdgendt) [reply drivers/flash/Kconfig.mspi:8]: I'd say it's a common way of solving things using `Kconfig`, and avoids summing up all possible drivers.

Comment (swift-tk) [reply drivers/flash/Kconfig.mspi:8]: I have an idea for generating common includes in bindings. 
We could generate a compat DT_HAS macro for those that has a compatible string in yaml.
In my case, with ambiq,mspi-device or mspi-atxp032 includes mspi-device and with compatible: "mspi-device" added, we then generate DT_HAS_* macro whenever we encouter `compatible` in the includes as well.

Comment (pdgendt) [reply drivers/flash/Kconfig.mspi:8]: Sure, but why overcomplicate things with the need for users to know all devicetree compatibles/kconfig options?

The way I see this, is that these are just flash drivers, and if they are enabled in devicetree and `CONFIG_FLASH` is enabled, they should be compiled in. Having the additional option to disable them with `CONFIG_FLASH_MSPI=n` is more counterintuitive IMO.

Comment (swift-tk) [reply drivers/flash/Kconfig.mspi:8]: > Sure, but why overcomplicate things with the need for users to know all devicetree compatibles/kconfig options?
> 
> The way I see this, is that these are just flash drivers, and if they are enabled in devicetree and `CONFIG_FLASH` is enabled, they should be compiled in. Having the additional option to disable them with `CONFIG_FLASH_MSPI=n` is more counterintuitive IMO.

oops, my bad, I should delete the prompt for config FLASH_MSPI. It is my intention for it to be hidden. My goal was to have a hierachical structure and avoid having to write all the same things to its members.

Comment (pdgendt) [reply drivers/flash/Kconfig.mspi:8]: > oops, my bad, I should delete the prompt for config FLASH_MSPI. It is my intention for it to be hidden. My goal was to have a hierachical structure and avoid having to write all the same things to its members.

The more reason to let the symbol be selected by the driver options instead of constructing `default y` and `depends on`.

By the way, the same should be done for `MEMC_MSPI`.

Comment (swift-tk) [reply drivers/flash/Kconfig.mspi:8]: I see your point. 
Generating common include DT_HAS macro would still litter...
I will make the change.

**Review by pdgendt — APPROVED** [review]
Nit: change the commit subject to something like

```
drivers: Hide MSPI Kconfig symbols and select from drivers
```

**Comment by pdgendt** [comment]
Off-topic question: are @Swift-T and @swift-tk different accounts from the same user? If so, you aren't supposed to self-approve PRs.

**Comment by Swift-T** [comment]
> Off-topic question: are @Swift-T and @swift-tk different accounts from the same user? If so, you aren't supposed to self-approve PRs.

@pdgendt  Yes, same guy. I'm just trying out the mspi file list in the MAINTAINERS.yml. Shall I assign you? or someone else?

**Comment by pdgendt** [comment]
> > Off-topic question: are @Swift-T and @swift-tk different accounts from the same user? If so, you aren't supposed to self-approve PRs.
> 
> @pdgendt Yes, same guy. I'm just trying out the mspi file list in the MAINTAINERS.yml. Shall I assign you? or someone else?

Let's assign to both accounts for now to make it clear, please create PRs from your member/assignee account next time.

**Comment by Swift-T** [comment]
> > > Off-topic question: are @Swift-T and @swift-tk different accounts from the same user? If so, you aren't supposed to self-approve PRs.
> > 
> > 
> > @pdgendt Yes, same guy. I'm just trying out the mspi file list in the MAINTAINERS.yml. Shall I assign you? or someone else?
> 
> Let's assign to both accounts for now to make it clear, please create PRs from your member/assignee account next time.

Don't PRs require assignee approval? Since I'm not supposed to approve myself?

**Review by nashif — APPROVED** [review]

**Comment by pdgendt** [comment]
> Don't PRs require assignee approval? Since I'm not supposed to approve myself?

https://merge-list.zephyrproject.io/ shows approval is accepted, it just needs to wait for the 48 hours review time.

**Comment by swift-tk** [comment]
> are @Swift-T and @swift-tk the same person?

yes

**Comment by nashif** [comment]
> > are @Swift-T and @swift-tk the same person?
> 
> yes

Can you please just use one GH handle when dealing with the project? It is confusing to see see a PR submitted by one handle and comments being sent using another.

**Comment by swift-tk** [comment]
> > > are @Swift-T and @swift-tk the same person?
> > 
> > 
> > yes
> 
> Can you please just use one GH handle when dealing with the project? It is confusing to see see a PR submitted by one handle and comments being sent using another.

Roger that, I’m just trying to see if the mspi file list is working and I can be added as reviewers by bot.

**Review by de-nordic — APPROVED** [review]

**Pull request merged**
